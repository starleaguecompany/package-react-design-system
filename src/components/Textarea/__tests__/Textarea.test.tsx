import * as React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Menu } from '@starleaguecompany/react-icons';
import { Textarea } from '../../../index';

describe('Textarea', () => {
  test('should render correctly with no props', () => {
    render(<Textarea />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByTestId('Textarea')).toHaveClass('container');
  });

  test('should render correctly with attributes', () => {
    render(<Textarea id="test-id" className="test-class" />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveAttribute('id', 'test-id');
    expect(screen.getByTestId('Textarea')).toHaveClass('test-class');
  });

  test('should render correctly with value prop', () => {
    render(<Textarea value="Test text" />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveValue('Test text');
    expect(screen.getByText('Test text')).toBeVisible();
  });

  test('should render correctly with disabled prop', () => {
    const { container } = render(<Textarea label="Test label" disabled />);
    const disabled = container.querySelectorAll('.disabled');

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByTestId('Textarea')).toHaveClass('disabled');
    expect(disabled.length).toEqual(3);
    expect(screen.getByRole('textbox')).toBeDisabled();
  });

  test('should render correctly with invalid prop', () => {
    const { container } = render(<Textarea label="Test label" invalid />);
    const invalid = container.querySelectorAll('.invalid');

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByTestId('Textarea')).toHaveClass('invalid');
    expect(screen.getByRole('textbox')).toHaveClass('invalid');
    expect(invalid.length).toEqual(3);
  });

  test('should render correctly with loading prop', () => {
    render(<Textarea loading />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByTestId('Textarea')).toContainElement(screen.getByTestId('Spinner'));
    expect(screen.getByRole('textbox')).toHaveClass('with-icon');
  });

  test('should render correctly with icon prop', () => {
    render(<Textarea icon={<Menu />} />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByTestId('Textarea')).toContainElement(screen.getByTestId('Menu'));
    expect(screen.getByRole('textbox')).toHaveClass('with-icon');
  });

  test('should render correctly with placeholder prop', () => {
    render(<Textarea placeholder="Test placeholder" />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByText('Test placeholder')).toBeVisible();
    expect(screen.getByText('Test placeholder')).toHaveClass('placeholder');
  });

  test('should render correctly with readOnly prop', () => {
    render(<Textarea readOnly value="Test value" />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveAttribute('readonly');
    userEvent.click(screen.getByRole('textbox'));
    userEvent.keyboard('New value');
    expect(screen.getByRole('textbox')).toHaveValue('Test value');
  });

  test('should render correctly with rows prop', () => {
    render(<Textarea rows={4} />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveAttribute('rows', '4');
  });

  test('should render correctly with resizable prop', () => {
    render(<Textarea resizable />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    expect(screen.getByRole('textbox')).toHaveClass('resize');
  });

  test('should execute onChange callback when typing', () => {
    const onChangeCallback = jest.fn();
    render(<Textarea value="Test value" onChange={onChangeCallback} />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    userEvent.click(screen.getByRole('textbox'));
    userEvent.keyboard('12345');
    expect(onChangeCallback).toHaveBeenCalledTimes(5);
  });

  test('should execute onFocus callback on click textarea', () => {
    const onFocusCallback = jest.fn();
    render(<Textarea onFocus={onFocusCallback} />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    userEvent.click(screen.getByRole('textbox'));
    expect(screen.getByRole('textbox')).toHaveFocus();
    expect(onFocusCallback).toHaveBeenCalledTimes(1);
  });

  test('should execute onBlur callback when blur focus', () => {
    const onBlurCallback = jest.fn();
    render(<Textarea onBlur={onBlurCallback} />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    userEvent.click(screen.getByRole('textbox'));
    userEvent.click(document.body);
    expect(screen.getByRole('textbox')).not.toHaveFocus();
    expect(onBlurCallback).toHaveBeenCalledTimes(1);
  });

  test('should render correctly on external value update', () => {
    const onChangeCallback = jest.fn();
    const { rerender } = render(<Textarea value="" onChange={onChangeCallback} />);

    expect(screen.getByTestId('Textarea')).toMatchSnapshot();
    rerender(<Textarea value="123" onChange={onChangeCallback} />);
    expect(screen.getByText('123')).toBeVisible();
    expect(screen.getByRole('textbox')).toHaveValue('123');
    expect(onChangeCallback).toBeCalledTimes(0);
  });
});
