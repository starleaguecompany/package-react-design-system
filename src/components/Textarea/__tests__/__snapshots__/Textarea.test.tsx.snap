// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should execute onBlur callback when blur focus 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    rows="1"
    spellcheck="false"
  />
</div>
`;

exports[`Textarea should execute onChange callback when typing 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea filled resize"
    rows="1"
    spellcheck="false"
  >
    Test value
  </textarea>
</div>
`;

exports[`Textarea should execute onFocus callback on click textarea 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    rows="1"
    spellcheck="false"
  />
</div>
`;

exports[`Textarea should render correctly on external value update 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    rows="1"
    spellcheck="false"
  />
</div>
`;

exports[`Textarea should render correctly with attributes 1`] = `
<div
  class="test-class container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    id="test-id"
    rows="1"
    spellcheck="false"
  />
</div>
`;

exports[`Textarea should render correctly with disabled prop 1`] = `
<div
  class="container disabled with-label"
  data-qa="Textarea"
>
  <textarea
    class="textarea disabled resize"
    disabled=""
    rows="1"
    spellcheck="false"
  />
  <label
    class="placeholder disabled"
  >
    Test label
  </label>
</div>
`;

exports[`Textarea should render correctly with icon prop 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea with-icon resize"
    rows="1"
    spellcheck="false"
  />
  <span
    class="icon container size-20 shape-round"
    data-qa="Icon"
  >
    <svg
      data-qa="Menu"
      fill="none"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 5h12M4 10h12M4 15h12"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
      />
    </svg>
  </span>
</div>
`;

exports[`Textarea should render correctly with invalid prop 1`] = `
<div
  class="container invalid with-label"
  data-qa="Textarea"
>
  <textarea
    class="textarea invalid resize"
    rows="1"
    spellcheck="false"
  />
  <label
    class="placeholder invalid"
  >
    Test label
  </label>
</div>
`;

exports[`Textarea should render correctly with loading prop 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea with-icon resize"
    rows="1"
    spellcheck="false"
  />
  <span
    class="icon container size-20"
    data-qa="Spinner"
  >
    <svg
      class="svg"
      viewBox="0 0 150 150"
    >
      <circle
        class="circle circle-1"
        cx="75"
        cy="75"
        r="60"
      />
      <circle
        class="circle circle-2"
        cx="75"
        cy="75"
        r="60"
      />
    </svg>
  </span>
</div>
`;

exports[`Textarea should render correctly with no props 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    rows="1"
    spellcheck="false"
  />
</div>
`;

exports[`Textarea should render correctly with placeholder prop 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    rows="1"
    spellcheck="false"
  />
  <label
    class="placeholder"
  >
    Test placeholder
  </label>
</div>
`;

exports[`Textarea should render correctly with readOnly prop 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea filled resize"
    readonly=""
    rows="1"
    spellcheck="false"
  >
    Test value
  </textarea>
</div>
`;

exports[`Textarea should render correctly with resizable prop 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    rows="1"
    spellcheck="false"
  />
</div>
`;

exports[`Textarea should render correctly with rows prop 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea resize"
    rows="4"
    spellcheck="false"
  />
</div>
`;

exports[`Textarea should render correctly with value prop 1`] = `
<div
  class="container"
  data-qa="Textarea"
>
  <textarea
    class="textarea filled resize"
    rows="1"
    spellcheck="false"
  >
    Test text
  </textarea>
</div>
`;
